validator-network/src/network_impl.rs

```diff
@@ -127,6 +127,19 @@ where
             Err(NetworkError::UnknownValidator(validator_id))
         }
     }
+
+    /// Clears the validator->peer_id cache.
+    /// The cached entry should be cleared when the peer id might have changed.
+    fn clear_validator_peer_id_cache(&self, validator_id: usize) {
+        let state = self.state.write();
+        let validator_key = state.validator_keys.get(validator_id);
+        if let Some(validator_key) = validator_key {
+            self.state
+                .write()
```

Comment:
Releasing and immediately acquiring the write lock again sounds wrong to me.

Reply:
Looks like a reentrant lock here, this would deadlock I believe as the write taken in L134 would not be dropped. I looked a bit at the logs produced by the reconnect test and they support that at least on the surface.

I think keep using the already taken state should fix it.